### PR TITLE
Support mixed domain assembly

### PR DESCRIPTION
**Pull request by jpdean**
This PR adds support for mixed domain assembly. Forms with test functions, trial functions, and coefficients defined over different meshes can be assembled. The user must provide a map to relate mesh entities in the integration domain to mesh entities in the mesh the function is defined over; this is best illustrated by the new tests.

The codimension 1 case requires [this](https://github.com/FEniCS/ffcx/tree/jpdean/mixed_dim) branch of FFCX. In a future PR, I will add permutations which are required in some cases.

There are still lots of parts of the code that need to be tidied up / refined. Any feedback on the implementation approach would be much appreciated.

### CODE REVIEW

File: cpp/dolfinx/fem/sparsitybuild.h [hunk 1]
```diff
@@ -36,10 +36,16 @@ namespace sparsitybuild
 /// over
 /// @param[in] dofmaps The dofmap to use in building the sparsity
 /// pattern
+/// @param[in] cell_maps Maps from the cells in the integration domain to
+/// the cells in each function space in the form.
```
Comment (chrisrichardson) [inline cpp/dolfinx/fem/sparsitybuild.h:40]: Add a description of the default.

File: cpp/dolfinx/fem/sparsitybuild.h [hunk 2]
```diff
@@ -48,9 +54,15 @@ void cells(
 /// @param[in] dofmaps The dofmap to use in building the sparsity
 /// pattern
 /// @note The sparsity pattern is not finalised
+/// @param[in] cell_maps Maps from the cells in the integration domain to
+/// the cells in each function space in the form.
```
Comment (chrisrichardson) [inline cpp/dolfinx/fem/sparsitybuild.h:58]: ditto: describe default

File: cpp/dolfinx/fem/Form.h [hunk 3]
```diff
@@ -102,18 +108,24 @@ class Form
...
     if (!_mesh and !function_spaces.empty())
       _mesh = function_spaces[0]->mesh();
     for (const auto& V : function_spaces)
     {
-      if (_mesh != V->mesh())
-        throw std::runtime_error("Incompatible mesh");
+      if (_mesh != V->mesh()
+          and entity_maps.find(V->mesh()) == entity_maps.end())
+        throw std::runtime_error(
+            "Incompatible mesh. entity_maps must be provided.");
```
Comment (chrisrichardson) [inline cpp/dolfinx/fem/Form.h:128]:
```suggestion
            "Incompatible mesh: entity_maps must be provided.");
```
Just a minor tweak

File: cpp/dolfinx/fem/assemble_matrix_impl.h [hunk 4]
```diff
@@ -80,6 +85,12 @@ void assemble_cells(
   for (std::size_t index = 0; index < cells.size(); ++index)
   {
     std::int32_t c = cells[index];
+    // Map the cell in the integration domain to the cell in the mesh
+    // each function space is defined over
+    std::int32_t c_0 = cell_map_0(cells.subspan(index, 1));
```
Comment (chrisrichardson) [inline cpp/dolfinx/fem/assemble_matrix_impl.h:90]: What is the point of `subspan` here? Isn't it just the same as `cells[index]`?

Comment (chrisrichardson) [reply cpp/dolfinx/fem/assemble_matrix_impl.h:90]: OK, I see that `cell_map_0` has to take a `span` argument... But if it is always length 1, does that make sense? I guess for facets, it is length 2, so it is consistent...

**Comment by garth-wells** [comment]
Closing for now, re-open later when ready.

**Pull request closed without merging**